File: java/src/test/java/org/mydotey/objectpool/threadpool/ObjectPoolTest.java
package org.mydotey.objectpool.threadpool;

import java.io.IOException;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.Assert;
import org.junit.Ignore;
import org.junit.Test;
import org.mydotey.objectpool.ObjectPool;
import org.mydotey.objectpool.facade.ThreadPools;
import org.mydotey.objectpool.threadpool.DefaultThreadPool;
import org.mydotey.objectpool.threadpool.ThreadPool;
import org.mydotey.objectpool.threadpool.ThreadPoolConfig;

/**
 * @author koqizhao
 *
 * Feb 6, 2018
 */
public class ObjectPoolTest {

    protected int _minSize = 10;
    protected int _maxSize = 100;
    protected int _submitterThreadCount = 10;

    protected int _defaultTaskCount = _minSize;
    protected long _defaultTaskSleep = 0;
    protected long _defaultViInitDelay = 1000;
    protected int _defaultSizeAfterSumit = -1;
    protected long _defaultFinishSleep = 0;
    protected int _defaultFinalSize = -1;

    protected ThreadPool newThreadPool() {
        ThreadPoolConfig.Builder builder = ThreadPools.newThreadPoolConfigBuilder();
        builder.setMinSize(_minSize).setMaxSize(_maxSize);
        return ThreadPools.newThreadPool(builder.build());
    }

    @Test
    public void threadPoolCreateTest() throws IOException {
        System.out.println();
        System.out.println();

        ThreadPool pool = newThreadPool();
        System.out.println("Pool Size: " + pool.getSize());
        Assert.assertEquals(_minSize, pool.getSize());
        pool.close();
    }

    @Test
    public void threadPoolSubmitTaskTest() throws IOException, InterruptedException {
        int taskCount = _minSize;
        long taskSleep = _defaultTaskSleep;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _minSize;
        long finishSleep = 1;
        int finalSize = _defaultFinalSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
    }

    @Test
    public void threadPoolSubmitTaskTest2() throws IOException, InterruptedException {
        int taskCount = _minSize;
        long taskSleep = 10;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _minSize;
        long finishSleep = 1000;
        int finalSize = _minSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
    }

    @Test
    public void threadPoolSubmitTaskTest3() throws IOException, InterruptedException {
        int taskCount = 50;
        long taskSleep = 50;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _defaultSizeAfterSumit;
        long finishSleep = 1000;
        int finalSize = _defaultFinalSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
    }

    @Test
    public void threadPoolSubmitTaskTest4() throws IOException, InterruptedException {
        int taskCount = _maxSize;
        long taskSleep = 500;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _maxSize;
        long finishSleep = 5000;
        int finalSize = _maxSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
    }

    @Test
    public void threadPoolSubmitTaskTest5() throws IOException, InterruptedException {
        int taskCount = 200;
        long taskSleep = 50;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _maxSize;
        long finishSleep = 1000;
        int finalSize = _maxSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
    }

    @Test
    public void threadPoolSubmitTaskTest6() throws IOException, InterruptedException {
        int taskCount = 200;
        long taskSleep = 2000;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _maxSize;
        long finishSleep = 5000;
        int finalSize = _maxSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
    }

    @Test
    @Ignore
    public void threadPoolSubmitTaskStressTest() throws IOException, InterruptedException {
        int count = 100;
        for (int i = 0; i < count; i++)
            threadPoolSubmitTaskTest();
    }

    @Test
    @Ignore
    public void threadPoolSubmitTaskStressTest2() throws IOException, InterruptedException {
        int count = 100;
        for (int i = 0; i < count; i++)
            threadPoolSubmitTaskTest4();
    }

    @Test
    public void threadPoolSubmitTaskConcurrentTest() throws IOException, InterruptedException {
        int taskCount = 200;
        long taskSleep = 2000;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _defaultSizeAfterSumit;
        long finishSleep = 5000;
        int finalSize = _maxSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                newThreadPool(), SubmissionMode.Concurrent);
    }

    @Test
    public void threadPoolSubmitTaskConcurrentTest2() throws IOException, InterruptedException {
        int taskCount = 200;
        long taskSleep = 2000;
        long viInitDelay = _defaultViInitDelay;
        int sizeAfterSubmit = _defaultSizeAfterSumit;
        long finishSleep = 5000;
        int finalSize = _maxSize;
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                newThreadPool(), SubmissionMode.SelfSelf);
    }

    protected void threadPoolSubmitTaskTest(int taskCount, long taskSleep, long viInitDelay, int sizeAfterSubmit,
            long finishSleep, int finalSize) throws IOException, InterruptedException {
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                newThreadPool());
    }

    protected void threadPoolSubmitTaskTest(int taskCount, long taskSleep, long viInitDelay, int sizeAfterSubmit,
            long finishSleep, int finalSize, ThreadPool customPool) throws IOException, InterruptedException {
        threadPoolSubmitTaskTest(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize, customPool,
                SubmissionMode.SingleThread);
    }

    protected void threadPoolSubmitTaskTest(int taskCount, long taskSleep, long viInitDelay, int sizeAfterSubmit,
            long finishSleep, int finalSize, ThreadPool customPool, SubmissionMode submissionMode)
            throws IOException, InterruptedException {
        System.out.println();
        System.out.println();

        AtomicInteger counter = new AtomicInteger();
        CountDownLatch countDownLatch = new CountDownLatch(taskCount);
        Runnable task = () -> {
            counter.incrementAndGet();
            countDownLatch.countDown();
            if (taskSleep <= 0)
                return;

            try {
                Thread.sleep(taskSleep);
            } catch (Exception e) {
            }
        };
        long now = System.currentTimeMillis();
        ExecutorService taskSubmitter = null;
        ScheduledExecutorService viTimer = null;
        ThreadPool self = null;
        try (ThreadPool pool = customPool) {
            System.out.println("new thread pool eclipsed: " + (System.currentTimeMillis() - now));
            System.out.println("counter value: " + counter);
            vi(pool);
            Assert.assertEquals(0, counter.get());

            viTimer = Executors.newSingleThreadScheduledExecutor();
            viTimer.scheduleWithFixedDelay(() -> {
                System.out.println("counter value: " + counter);
                vi(pool);
            }, 1000, 500, TimeUnit.MILLISECONDS);

            switch (submissionMode) {
                case SingleThread:
                    for (int i = 0; i < taskCount; i++)
                        pool.submit(task);
                    break;
                case Concurrent:
                    taskSubmitter = Executors.newFixedThreadPool(_submitterThreadCount);
                    now = System.currentTimeMillis();
                    for (int i = 0; i < taskCount; i++)
                        taskSubmitter.submit(() -> {
                            try {
                                pool.submit(task);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        });
                    break;
                case SelfSelf:
                    self = newThreadPool();
                    for (int i = 0; i < taskCount; i++)
                        self.submit(() -> {
                            try {
                                pool.submit(task);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        });
                    break;
                default:
                    Assert.fail("impossible submissionMode: " + submissionMode);
                    break;
            }

            System.out.println("submit tasks eclipsed: " + (System.currentTimeMillis() - now));
            System.out.println("counter value: " + counter);
            vi(pool);

            Assert.assertTrue(_minSize <= pool.getSize());
            Assert.assertTrue(_maxSize >= pool.getSize());
            if (sizeAfterSubmit >= 0)
                Assert.assertEquals(sizeAfterSubmit, pool.getSize());

            countDownLatch.await();

            System.out.println("tasks run time: " + (System.currentTimeMillis() - now));
            System.out.println("counter value: " + counter);
            vi(pool);

            Assert.assertEquals(taskCount, counter.get());
            Assert.assertTrue(_minSize <= pool.getSize());
            Assert.assertTrue(_maxSize >= pool.getSize());

            if (finishSleep <= 0)
                return;

            Thread.sleep(finishSleep);

            vi(pool);

            Assert.assertEquals(taskCount, counter.get());
            Assert.assertTrue(_minSize <= pool.getSize());
            Assert.assertTrue(_maxSize >= pool.getSize());

            if (finalSize >= 0)
                Assert.assertEquals(finalSize, pool.getSize());
        } finally {
            if (taskSubmitter != null)
                taskSubmitter.shutdown();

            if (self != null)
                self.close();

            if (viTimer != null)
                viTimer.shutdown();
        }
    }

    protected void vi(ThreadPool pool) {
        System.out.println();
        ObjectPool<?> objectPool = ((DefaultThreadPool) pool).getObjectPool();
        System.out.println("pool size: " + objectPool.getSize());
        System.out.println("pool acquired size: " + objectPool.getAcquiredSize());
        System.out.println("pool available size: " + objectPool.getAvailableSize());
        System.out.println();
    }

    protected enum SubmissionMode {
        SingleThread, Concurrent, SelfSelf
    }
}

File: dotnet/test/MyDotey.ObjectPool.Tests/ObjectPoolTest.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Xunit;
using NLog;
using NLog.Config;
using NLog.Targets;
using MyDotey.ObjectPool.Facade;
using MyDotey.ObjectPool.ThreadPool;

/**
 * @author koqizhao
 *
 * Feb 25, 2018
 */
namespace MyDotey.ObjectPool.Tests
{
    public class ObjectPoolTest
    {
        protected internal static long CurrentTimeMillis { get { return DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond; } }

        protected int _minSize = 10;
        protected int _maxSize = 100;
        protected int _submitterThreadCount = 10;

        protected int _defaultTaskCount = 10;
        protected long _defaultTaskSleep = 0;
        protected long _defaultViInitDelay = 1000;
        protected int _defaultSizeAfterSubmit = -1;
        protected long _defaultFinishSleep = 0;
        protected int _defaultFinalSize = -1;

        public ObjectPoolTest()
        {
            var config = new LoggingConfiguration();
            var logconsole = new ConsoleTarget() { Name = "logconsole" };
            config.LoggingRules.Add(new NLog.Config.LoggingRule("*", LogLevel.Trace, logconsole));
            NLog.LogManager.Configuration = config;
        }

        protected virtual IThreadPool NewThreadPool()
        {
            return NewThreadPool(0);
        }

        protected virtual IThreadPool NewThreadPool(int queueCapacity)
        {
            IBuilder Builder = ThreadPools.NewThreadPoolConfigBuilder();
            Builder.SetMinSize(_minSize).SetMaxSize(_maxSize).SetQueueCapacity(queueCapacity);
            return ThreadPools.NewThreadPool(Builder.Build());
        }

        [Fact]
        public virtual void ThreadPoolCreateTest()
        {
            Console.WriteLine();
            Console.WriteLine();

            IThreadPool pool = NewThreadPool();
            Console.WriteLine("Pool Size: " + pool.Size);
            Assert.Equal(_minSize, pool.Size);
            pool.Dispose();
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest()
        {
            int taskCount = _minSize;
            long taskSleep = _defaultTaskSleep;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _minSize;
            long finishSleep = 1;
            int finalSize = _defaultFinalSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest2()
        {
            int taskCount = _minSize;
            long taskSleep = 10;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _minSize;
            long finishSleep = 1000;
            int finalSize = _minSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest3()
        {
            int taskCount = 50;
            long taskSleep = 100;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _defaultSizeAfterSubmit;
            long finishSleep = 1000;
            int finalSize = _defaultFinalSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest4()
        {
            int taskCount = _maxSize;
            long taskSleep = 500;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _maxSize;
            long finishSleep = 5000;
            int finalSize = _maxSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest5()
        {
            int taskCount = 200;
            long taskSleep = 100;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _maxSize;
            long finishSleep = 1000;
            int finalSize = _maxSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest6()
        {
            int taskCount = 200;
            long taskSleep = 2000;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _maxSize;
            long finishSleep = 5000;
            int finalSize = _maxSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskTest7()
        {
            int taskCount = 200;
            long taskSleep = 2000;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _maxSize;
            long finishSleep = 5000;
            int finalSize = _maxSize;
            int queueCapacity = 10;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                NewThreadPool(queueCapacity));
        }

        [Fact(Skip = "stress")]
        public virtual void ThreadPoolSubmitTaskStressTest()
        {
            int count = 100;
            for (int i = 0; i < count; i++)
                ThreadPoolSubmitTaskTest();
        }

        [Fact(Skip = "stress")]
        public virtual void ThreadPoolSubmitTaskStressTest2()
        {
            int count = 100;
            for (int i = 0; i < count; i++)
                ThreadPoolSubmitTaskTest4();
        }

        [Fact(Skip = "stress")]
        public virtual void ThreadPoolSubmitTaskStressTest3()
        {
            int count = 100;
            for (int i = 0; i < count; i++)
                ThreadPoolSubmitTaskTest7();
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskConcurrentTest()
        {
            int taskCount = 200;
            long taskSleep = 2000;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _defaultSizeAfterSubmit;
            long finishSleep = 5000;
            int finalSize = _maxSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                    NewThreadPool(), SubmissionMode.Concurrent);
        }

        [Fact]
        public virtual void ThreadPoolSubmitTaskConcurrentTest2()
        {
            int taskCount = 200;
            long taskSleep = 2000;
            long viInitDelay = _defaultViInitDelay;
            int sizeAfterSubmit = _defaultSizeAfterSubmit;
            long finishSleep = 5000;
            int finalSize = _maxSize;
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                    NewThreadPool(), SubmissionMode.SelfSelf);
        }

        protected virtual void ThreadPoolSubmitTaskTestInternal(int taskCount, long taskSleep, long viInitDelay, int sizeAfterSubmit,
                long finishSleep, int finalSize)
        {
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize,
                    NewThreadPool());
        }

        protected virtual void ThreadPoolSubmitTaskTestInternal(int taskCount, long taskSleep, long viInitDelay, int sizeAfterSubmit,
                long finishSleep, int finalSize, IThreadPool customPool)
        {
            ThreadPoolSubmitTaskTestInternal(taskCount, taskSleep, viInitDelay, sizeAfterSubmit, finishSleep, finalSize, customPool,
                    SubmissionMode.SingleThread);
        }

        protected virtual void ThreadPoolSubmitTaskTestInternal(int taskCount, long taskSleep, long viInitDelay, int sizeAfterSubmit,
                long finishSleep, int finalSize, IThreadPool customPool, SubmissionMode submissionMode)
        {
            Console.WriteLine();
            Console.WriteLine();

            int counter = 0;
            CountdownEvent countDownLatch = new CountdownEvent(taskCount);
            Action task = () =>
            {
                Interlocked.Increment(ref counter);
                countDownLatch.Signal();
                if (taskSleep <= 0)
                    return;

                try
                {
                    Thread.Sleep((int)taskSleep);
                }
                catch
                {
                }
            };
            long now = CurrentTimeMillis;
            Timer ViTimer = null;
            IThreadPool self = null;
            using (IThreadPool pool = customPool)
            {
                Console.WriteLine("new thread pool eclipsed: " + (CurrentTimeMillis - now));
                Console.WriteLine("counter value: " + counter);
                Vi(pool);
                Assert.Equal(0, counter);

                using (ViTimer = new Timer(s =>
                 {
                     Console.WriteLine("counter value: " + counter);
                     Vi(pool);
                 }, null, 1000, 500))
                {
                    switch (submissionMode)
                    {
                        case SubmissionMode.SingleThread:
                            for (int i = 0; i < taskCount; i++)
                                pool.Submit(task);
                            break;
                        case SubmissionMode.Concurrent:
                            now = CurrentTimeMillis;
                            for (int i = 0; i < taskCount; i++)
                                Task.Run(() =>
                                {
                                    try
                                    {
                                        pool.Submit(task);
                                    }
                                    catch (ThreadInterruptedException e)
                                    {
                                        Console.WriteLine(e);
                                    }
                                });
                            break;
                        case SubmissionMode.SelfSelf:
                            self = NewThreadPool();
                            for (int i = 0; i < taskCount; i++)
                                self.Submit(() =>
                                {
                                    try
                                    {
                                        pool.Submit(task);
                                    }
                                    catch (ThreadInterruptedException e)
                                    {
                                        Console.WriteLine(e);
                                    }
                                });
                            break;
                        default:
                            throw new Exception("impossible submissionMode: " + submissionMode);
                    }

                    Console.WriteLine("Submit tasks eclipsed: " + (CurrentTimeMillis - now));
                    Console.WriteLine("counter value: " + counter);
                    Vi(pool);

                    Assert.True(_minSize <= pool.Size);
                    Assert.True(_maxSize >= pool.Size);
                    if (sizeAfterSubmit >= 0)
                        Assert.Equal(sizeAfterSubmit, pool.Size);

                    countDownLatch.Wait();

                    Console.WriteLine("tasks run time: " + (CurrentTimeMillis - now));
                    Console.WriteLine("counter value: " + counter);
                    Vi(pool);

                    Assert.Equal(taskCount, counter);
                    Assert.True(_minSize <= pool.Size);
                    Assert.True(_maxSize >= pool.Size);

                    if (finishSleep <= 0)
                        return;

                    Thread.Sleep((int)finishSleep);

                    Vi(pool);

                    Assert.Equal(taskCount, counter);
                    Assert.True(_minSize <= pool.Size);
                    Assert.True(_maxSize >= pool.Size);

                    if (finalSize >= 0)
                        Assert.Equal(finalSize, pool.Size);
                }
            }
        }

        protected virtual void Vi(IThreadPool pool)
        {
            Console.WriteLine();
            IObjectPool<WorkerThread> objectPool = ((DefaultThreadPool)pool).ObjectPool;
            Console.WriteLine("pool size: " + objectPool.Size);
            Console.WriteLine("pool acquired size: " + objectPool.AcquiredSize);
            Console.WriteLine("pool available size: " + objectPool.AvailableSize);
            Console.WriteLine();
        }

        protected enum SubmissionMode
        {
            SingleThread, Concurrent, SelfSelf
        }
    }
}
